README.rst

simplerouter
============

simplerouter is a simple WSGI/WebOb router partially based on
the router described in `WebOb's DIY Framework Tutorial
<http://docs.webob.org/en/latest/do-it-yourself.html>`_.
Python 2.7 and Python 3.2 and newer are supported.

Documentation is available at `readthedocs.org
<http://simplerouter.readthedocs.org/en/latest/>`_.

The main git repository is available at `Bitbucket
<https://bitbucket.org/rschoon/simplerouter>`_.


Installing
----------

simplerouter uses a setup.py script in the usual fashion, like so::

    $ python ./setup.py install

Alternately, simplerouter is available on pypi and can be installed
using pip::

    $ pip install simplerouter


Quick Example
-------------

app.py:

.. code-block:: python

    from simplerouter import Router

    router = Router()
    # view names are composed of modulename:function
    router.add_route('/post/{name}', 'views:post_view')
    router.add_route('/', 'views:index_view')

    application = router.as_wsgi

    if __name__=='__main__':
        from wsgiref.simple_server import make_server
        make_server('', 8000, application).serve_forever()

views.py:

.. code-block:: python

    from webob import Response

    def post_view(request):
        post_name = request.urlvars['name']
        # ... process post_name
        return Response("Post output for %s"%post_name)

    def index_view(request):
        return Response("Site index")


CHANGES.rst

Change Log
==========

1.2 (May 2 2015)
---------

* Add ``Router.reverse`` method for constructing URLs from view parameters.
* Allow views to be limited to specific HTTP method names.

1.1.1 (May 27 2014)
-------------------

* Packaging fixes
* Update tests for newer versions of Python

1.1 (Dec 1 2013)
----------------

* Improve support for nesting Router objects.
* Catch raised error responses.
* Improve test coverage

1.0 (Nov 10 2013)
-----------------

* First initial release.


tests.py

from webob import Request, Response, exc
from nose.tools import eq_, raises

class view_factory(object):
    def __init__(self, payload):
        self.payload = payload

    def __call__(self, request):
        if len(request.urlvars) > 0:
            return self.payload, request.urlvars
        return self.payload

#
# Simple Tests
#

def test_func_view():
    from simplerouter import Router

    def view(request):
        return "hello!"

    r = Router()
    r.add_route('/test', view)

    assert r(Request.blank('/test')) == "hello!"

def test_simple_paths():
    from simplerouter import Router

    r = Router()
    r.add_route('/', view_factory('root'))
    r.add_route('/path', view_factory('path'))
    r.add_route('/path/', view_factory('path_slash'))
    r.add_route('/path/{element}', view_factory('path_var'))

    eq_(r(Request.blank('/')), "root")
    eq_(r(Request.blank('/path')), "path")
    eq_(r(Request.blank('/path/')), "path_slash")
    eq_(r(Request.blank('/path/pie')) , ("path_var", {'element' : 'pie'}))

def test_method():
    from simplerouter import Router

    r = Router()
    r.add_route('/', view_factory('root'))
    r.add_route('/path_str', view_factory('get'), method="GET")
    r.add_route('/path_str', view_factory('post'), method="POST")
    r.add_route('/path_str', view_factory('head'), method="HEAD")
    r.add_route('/path_str', view_factory('else'))
    r.add_route('/path_list', view_factory('get-post'), method=("GET", "POST"))

    eq_(r(Request.blank('/', POST={})), "root")
    eq_(r(Request.blank('/path_str')), "get")
    eq_(r(Request.blank('/path_str', POST={})), "post")
    eq_(r(Request.blank('/path_list')), "get-post")
    eq_(r(Request.blank('/path_list', POST={})), "get-post")
    eq_(r(Request.blank('/path_list', method="HEAD")), "get-post")

def test_default_default():
    from simplerouter import Router

    r = Router()
    eq_(r(Request.blank('/')).status_code, 404)
    eq_(r(Request.blank('/path')).status_code, 404)

def test_routes_arg():
    from simplerouter import Router

    r = Router(
        ('/', view_factory('root')),
        ('/path', view_factory('path')),
        ('/vars', view_factory('vars'), { 'vars' : { 'key' : 'value' } })
    )

    eq_(r(Request.blank('/')), "root")
    eq_(r(Request.blank('/path')), "path")
    eq_(r(Request.blank('/vars')), ('vars', {'key':'value'}))

def test_exception():
    from simplerouter import Router
    from webob.exc import HTTPTemporaryRedirect

    def view(req):
        raise HTTPTemporaryRedirect(location='/home')

    r = Router()
    r.add_route('/', view)

    eq_(r(Request.blank('/')).status_code, 307)

@raises(Exception)
def test_exception_nocatch():
    from simplerouter import Router
    from webob.exc import HTTPTemporaryRedirect

    def view(req):
        raise HTTPTemporaryRedirect(location='/home')

    r = Router(catch_raised_responses=False)
    r.add_route('/', view)

    r(Request.blank('/'))

def test_bad_url():
    from simplerouter import Router
    from webob.exc import HTTPBadRequest

    r = Router()
    r.add_route('/', view_factory('root'))

    req = Request.blank('/')
    req.environ['PATH_INFO'] = b'/\xef'.decode('latin-1')
    eq_(r(req).status_code, 400)

#
# WSGI Tests
#

def test_as_wsgi():
    from simplerouter import Router

    def call_app(app, path):
        return Request.blank(path).call_application(app)

    r = Router(default=None)
    r.add_route('/', lambda req: Response("root"))
    r.add_route('/path', lambda req: Response("path"))

    statusRoot, headersRoot, bodyRoot = call_app(r.as_wsgi, '/')
    assert statusRoot.startswith('200')
    eq_(b''.join(bodyRoot), b'root')

    statusRoot, headersRoot, bodyRoot = call_app(r.as_wsgi, '/path')
    assert statusRoot.startswith('200')
    eq_(b''.join(bodyRoot), b'path')

    statusInvalid, headersInvalid, bodyInvalid = call_app(r.as_wsgi, '/invalid')
    assert statusInvalid.startswith('500')

def test_wsgi_view():
    from simplerouter import Router

    def call_app(app, path):
        return Request.blank(path).call_application(app)

    def app(environ, start_response):
        start_response('200 OK', [('Content-Type', 'text/plain')])
        return [environ['PATH_INFO'].encode('utf-8')]

    r = Router()
    r.add_route('/', view_factory('root'))
    r.add_route('/app', app, wsgi=True)

    statusAppRoot, headersAppRoot, bodyAppRoot = call_app(r.as_wsgi, '/app/')
    assert statusAppRoot.startswith('200')
    eq_(b''.join(bodyAppRoot), b'/')
    
    statusAppRoot, headersAppRoot, bodyAppRoot = call_app(r.as_wsgi, '/app/sub')
    assert statusAppRoot.startswith('200')
    eq_(b''.join(bodyAppRoot), b'/sub')

#
# Regexes
#

def test_null():
    from simplerouter import Router

    r = Router()
    r.add_route(None, view_factory('catchall'))
    r.add_route('/', view_factory('something'), priority=1)

    eq_(r(Request.blank('/')), 'something')
    eq_(r(Request.blank('/x')), 'catchall')

def test_regex():
    from simplerouter import Router

    r = Router()
    r.add_route('/{d:\d+}', view_factory('digit'))
    r.add_route('/{d:\d+}/', view_factory('digitSlash'))
    r.add_route('/term/{t:[^_]+}', view_factory('incSlash'))

    eq_(r(Request.blank('/1234')), ('digit', {'d' : '1234'}))
    eq_(r(Request.blank('/1234/')), ('digitSlash', {'d' : '1234'}))
    eq_(r(Request.blank('/term/abc/def')), ('incSlash', {'t' : 'abc/def'}))

#
# Try Slash Tests
#

def test_wrong_slash():
    from simplerouter import Router

    r = Router(default=view_factory('wrong'))
    r.add_route('/path', view_factory('path'))

    r2 = Router(default=view_factory('wrong'))
    r2.add_route('/path/', view_factory('path'))

    eq_(r(Request.blank('/path')), "path")
    eq_(r2(Request.blank('/path/')), "path")
    eq_(r(Request.blank('/path/')), "wrong")
    eq_(r2(Request.blank('/path')), "wrong")

def test_wrong_slash_try_slashes():
    from simplerouter import Router

    r = Router(default=view_factory('wrong'), try_slashes=True)
    r.add_route('/path', view_factory('path'))

    r2 = Router(default=view_factory('wrong'), try_slashes=True)
    r2.add_route('/path/', view_factory('path'))

    eq_(r(Request.blank('/path')), "path")
    eq_(r2(Request.blank('/path/')), "path")

    respRedir = r(Request.blank('/path/'))
    eq_(respRedir.status_code, 307)
    eq_(respRedir.location, "http://localhost/path")

    respRedir = r2(Request.blank('/path'))
    eq_(respRedir.status_code, 307)
    eq_(respRedir.location, "http://localhost/path/")

def test_no_try_slash():
    from simplerouter import Router

    r = Router(try_slashes=True)
    r.add_route('/path', view_factory('path'), no_alt_redir=True)
    
    eq_(r(Request.blank('/path/')).status_code, 404)

#
# Priority Tests
#

def test_priority():
    from simplerouter import Router

    r = Router()
    r.add_route('/1', view_factory('low'), priority=-1)
    r.add_route('/1', view_factory('normal'))
    eq_(r(Request.blank('/1')), "normal")

    r.add_route('/2', view_factory('normal'))
    r.add_route('/2', view_factory('low'), priority=-1)
    eq_(r(Request.blank('/2')), "normal")

    r = Router()
    r.add_route('/3', view_factory('high'), priority=1)
    r.add_route('/3', view_factory('normal'))
    r.add_route('/3', view_factory('low'), priority=-1)
    eq_(r(Request.blank('/3')), "high")

#
# View lookup tests
#

def test_failed_lookup():
    from simplerouter import Router

    r = Router()
    r.add_route('/missingModule', 'nonexistent.invalid:view')
    r.add_route('/missingView', 'simplerouter:xxx_nonexistent_view')

    resp = r(Request.blank('/missingModule'))
    eq_(resp.status_code, 500)
    
    resp = r(Request.blank('/missingView'))
    eq_(resp.status_code, 500)

def test_named():
    from simplerouter import Router

    r = Router()
    r.add_route('/blank', 'simplerouter:blank_view')
    eq_(r(Request.blank('/blank')).body, b'')

def test_default_named():
    from simplerouter import Router

    r = Router(default='simplerouter:blank_view')
    eq_(r(Request.blank('/')).body, b'')

#
# Path adjustment tests
#

def test_path_info():
    from simplerouter import Router

    def view(req):
        return req.script_name, req.path_info
    
    r = Router()
    r.add_route('/test', view, path_info=True)
    r.add_route('/slash/', view, path_info='.*') # not a good path_info
    
    eq_(r(Request.blank('/test/pants')), ('/test', '/pants'))
    eq_(r(Request.blank('/slash/test')), ('/slash/', 'test'))
    eq_(r(Request.blank('/test')).status_code, 404)

def test_path_info_none():
    from simplerouter import Router

    def view(req):
        return req.script_name, req.path_info

    r = Router()
    r.add_route('/test', view, path_info=True)
    r.add_route(None, view, path_info=True)

    eq_(r(Request.blank('/test/pants')), ('/test', '/pants'))
    eq_(r(Request.blank('/test')), ('', '/test'))

def test_path_info_restore():
    from simplerouter import Router

    def nullview(req):
        return None

    def view(req):
        return req.script_name, req.path_info

    r = Router()
    r.add_route('/test', nullview, path_info=True, priority=100)
    r.add_route('/test/2', view)
    
    eq_(r(Request.blank('/test/2')), ('', '/test/2'))

#
# Nested
#

def test_nested():
    from simplerouter import Router

    r = Router(
        ('/test', Router(
            ('/test2', view_factory('test_test2')),
        ), { 'path_info' : True })
    )

    eq_(r(Request.blank('/test/test2')), 'test_test2')

def test_nested_shorthand():
    from simplerouter import Router

    r = Router(
        ('/test', [
            ('/test2', view_factory('test_test2')),
            ('/test3', view_factory('test_test3'), {'vars' : {'name' : 'value'}}),
        ], { 'path_info' : True })
    )

    eq_(r(Request.blank('/test/test2')), 'test_test2')
    eq_(r(Request.blank('/test/test3')), ('test_test3', {'name' : 'value'}))

def test_nested_shorthand_opt():
    from simplerouter import Router

    r = Router(
        ('/test', [
            ('/test2', view_factory('test_test2')),
            { 'default' : view_factory('default') }
        ], { 'path_info' : True })
    )

    eq_(r(Request.blank('/test/test2')), 'test_test2')
    eq_(r(Request.blank('/test/doesnotexist')), 'default')

def test_urlvars_restore():
    from simplerouter import Router

    def nullview(req):
        return None

    def return_first_urlvar(req):
        return req.urlvars['first']

    r = Router(
        ('/{first}', Router(
            ('/{second}', nullview),
        default=return_first_urlvar), { 'path_info' : True }),
    )

    eq_(r(Request.blank('/var1/var2')), 'var1')

#
# reverse
#

def test_reverse():
    from simplerouter import Router

    root_view = view_factory('root')
    path_view = view_factory('path')
    path_slash_view = view_factory('path_slash')
    path_var_view = view_factory('path_var')
    path_info_view = view_factory('path_info')

    r = Router()
    r.add_route('/', root_view)
    r.add_route('/path', path_view)
    r.add_route('/path/', path_slash_view)
    r.add_route('/path/{element}', path_var_view)
    r.add_route('/pi/{a}', path_info_view, path_info=True)
    r.add_route('/blank', 'simplerouter:blank_view')

    eq_(r.reverse(root_view), '/')
    eq_(r.reverse(path_view), '/path')
    eq_(r.reverse(path_slash_view), '/path/')
    eq_(r.reverse(path_var_view, {'element' : 'pie'}), '/path/pie')
    eq_(r.reverse(path_info_view, {'a' : 'apple'}, '/and/banana'), '/pi/apple/and/banana')
    eq_(r.reverse('simplerouter:blank_view'), '/blank')

@raises(ValueError)
def test_reverse_fail():
    from simplerouter import Router

    r = Router()
    r.add_route('/blank', 'simplerouter:blank_view')

    r.reverse('/doesnotexist')

@raises(TypeError)
def test_reverse_fail_type():
    from simplerouter import Router

    r = Router()
    r.reverse(object())

@raises(ValueError)
def test_reverse_fail_impossible():
    from simplerouter import Router

    r = Router()
    r.add_route(None, 'simplerouter:blank_view')
    r.reverse('simplerouter:blank_view')


simplerouter.py
"""
simplerouter is an expansion of DIY framework described in the
WebOB documentation.
    
"""

__version__ = '1.2'
__all__ = ['Router', 'lookup_view']

import sys
import re
from webob import exc, Request, Response

def blank_view(request):
    return Response()

def internal_error_view(msg):
    return lambda req: exc.HTTPInternalServerError(msg)

def not_found_view(request):
    return exc.HTTPNotFound()

PATH_INFO_VAR = '__path_info__'
VAR_REGEX = re.compile(r'{(\w+)(?::([^}]+))?\}')
def parse_template(template, path_info):
    fmt = []
    regex = []
    last_pos = 0

    for match in VAR_REGEX.finditer(template):
        regex.append(re.escape(template[last_pos:match.start()]))
        fmt.append(template[last_pos:match.start()])

        regex.append('(?P<%s>%s)' % (match.group(1), match.group(2) or '[^/]+'))
        fmt.append('{%s}'%(match.group(1), ))

        last_pos = match.end()

    regex.append(re.escape(template[last_pos:]))
    fmt.append(template[last_pos:])

    if path_info is not None:
        if path_info is True:
            path_info = '/.*'
        regex.append('(?P<%s>%s)' % (PATH_INFO_VAR, path_info))

    return re.compile('^%s$' % "".join(regex)), "".join(fmt)

def lookup_view(view):
    if callable(view):
        return view

    module_name, func_name = view.split(':', 1)
    try:
        __import__(module_name)
    except ImportError:
        return internal_error_view("Module %s does not exist."%module_name)

    module = sys.modules[module_name]
    try:
        return getattr(module, func_name)
    except AttributeError:
        return internal_error_view("Function %s not found on module %s"%(func_name, module_name))

class Route(object):
    def __init__(self, path_re, viewname, vars=None, wsgi=False, no_alt_redir=False, priority=0, path_info=None, method=None):
        if wsgi and path_info is None:
            path_info = True

        if path_re is not None or path_info is not None:
            if path_re is None:
                path_re = ""
            self.path_re, self.path_fmt = parse_template(path_re, path_info)
        else:
            self.path_fmt = None
            self.path_re = re.compile("")

        if callable(viewname):
            self._view = viewname
            if hasattr(self._view, "__name__"):
                self.viewname = self._view.__name__
            elif hasattr(self._view, "__class__"):
                self.viewname = self._view.__class__.__name__
        else:
            self.viewname = viewname

        self.vars = vars
        self.wsgi = wsgi
        self.no_alt_redir = no_alt_redir
        self.priority = priority

        if isinstance(method, str):
            self.method = (method, )
        else:
            self.method = method
        if self.method is not None and "GET" in self.method:
            self.method = list(self.method) + ["HEAD"]

    def __repr__(self):
        if self.method is None:
            method = ""
        else:
            method = "|".join(self.method)

        return "<Route(%s%s @ %s)>"%(method, self.viewname, self.path_re.pattern)

    def match(self, request, alt=False):
        if alt and self.no_alt_redir:
            return False
        if self.method is not None and request.method not in self.method:
            return False
        return self.path_re.match(request.path_info)

    @property
    def view(self):
        try:
            return self._view
        except AttributeError:
            pass
        view = lookup_view(self.viewname)
        self._view = view
        return view

    def __call__(self, request):
        m = self.match(request)
        if m is not None:
            script_name_orig = request.script_name
            path_info_orig = request.path_info
            urlvars_orig = request.urlvars

            urlvars = m.groupdict()
            if PATH_INFO_VAR in urlvars:
                del urlvars[PATH_INFO_VAR]
                begin, end = m.span(PATH_INFO_VAR)
                request.script_name += request.path_info[:begin]
                request.path_info = request.path_info[begin:end]

            request.urlvars = urlvars
            if self.vars is not None:
                request.urlvars.update(self.vars)

            if self.wsgi:
                return self.view
            else:
                resp = self.view(request)
                if resp is None:
                    request.script_name = script_name_orig
                    request.path_info = path_info_orig
                    request.urlvars = urlvars_orig
                return resp

class Router(object):
    def __init__(self, *routes, **options):
        self._set_options(**options)

        self.routes = []
        for route in routes:
            if isinstance(route[-1], dict):
                self.add_route(*route[:-1], **route[-1])
            else:
                self.add_route(*route)

    def _set_options(self, default=not_found_view, try_slashes=False, catch_raised_responses=True):
        if default is not None:
            self.default = lookup_view(default)
        else:
            self.default = None
        self.try_slashes = try_slashes
        self.catch_raised_responses = catch_raised_responses

    def add_route(self, path, view, **kwargs):
        """Add a route to the router."""
        if isinstance(view, (list, tuple)):
            if isinstance(view[-1], dict):
                view = Router(*view[:-1], **view[-1])
            else:
                view = Router(*view)

        route = Route(path, view, **kwargs)
        for i, rti in enumerate(self.routes):
            if rti.priority < route.priority:
                self.routes.insert(i, route)
                return
        
        self.routes.append(route)
        
    def __call__(self, req):
        """Invoke router as a view."""

        # verify url was decoded properly
        try:
            req.path_info, req.script_name
        except (UnicodeDecodeError, UnicodeEncodeError):
            return exc.HTTPBadRequest()

        # try normal view
        matches = set()
        for view in self.matches(req):
            if view is not None:
                try:
                    r = view(req)
                except exc.HTTPException as respexc:
                    if not self.catch_raised_responses:
                        raise
                    return respexc
                if r is not None:
                    return r
                matches.add(view)
        
        # try redirect to alternate path
        if self.try_slashes:
            if req.environ['PATH_INFO'].endswith("/"):
                req.environ['PATH_INFO'] = req.environ['PATH_INFO'][:-1]
            else:
                req.environ['PATH_INFO'] += "/"
            
            altView = self.match(req, True)
            if altView is not None and altView not in matches:
                return exc.HTTPTemporaryRedirect(location=req.url)
        
        if self.default is not None: 
            return self.default(req)
        
    def match(self, req, alt=False):
        """Return the first view that the given request matches."""
        for m in self.matches(req, alt):
            return m 

    def matches(self, req, alt=False):
        """Iterate through all views that the given request matches."""
        for route in self.routes:
            m = route.match(req, alt=alt)
            if m:
                if callable(m):
                    yield m
                else:
                    yield route

    def _find_route_by_identifier(self, route):
        """Find a route by its name or callable or itself."""
        if isinstance(route, Route):
            return route
        elif isinstance(route, str):
            for r in self.routes:
                if r.viewname == route:
                    return r
        elif callable(route):
            for r in self.routes:
                if r.view == route:
                    return r
        else:
            raise TypeError("Expected a string or route callable, but got `%s' instead", type(route).__name__)
        raise ValueError("No such route %r"%(route, ))

    def reverse(self, route, vars={}, path_info=None):
        route = self._find_route_by_identifier(route)
        if route.path_fmt is None:
            raise ValueError("%r cannot be reversed"%(route, ))

        url = route.path_fmt.format(**vars)
        if path_info is not None:
            url += path_info
        return url

    def as_wsgi(self, environ, start_response):
        """Invoke router as an wsgi application."""
        req = Request(environ)
        resp = self(req)
        if resp is None:
            start_response('500 Internal Server Error', [('Content-Type', 'text/plain')])
            return [b'no default in wsgi call']
        return resp(environ, start_response)

